File: assignments/assignment-2-solution/src-csharp/TeaShop.UnitTest/Domain/Query/QueryTests.cs
using Assignment2Solution.Domain.Inventory;
using Assignment2Solution.Domain.Query;
using Assignment2Solution.Domain.Query.Filters;
using Assignment2Solution.Domain.Query.Sorts;
using Moq;

namespace TeaShop.UnitTest.Domain.Query;

public class QueryTests
{
    private readonly List<InventoryItem> _testItems;

    public QueryTests()
    {
        _testItems = new List<InventoryItem>
        {
            new(Guid.NewGuid(), "Green Tea", 15.00m, 10, new StarRating(4)),
            new(Guid.NewGuid(), "Black Tea", 10.00m, 0, new StarRating(5)),
            new(Guid.NewGuid(), "Oolong Tea", 20.00m, 5, new StarRating(3)),
            new(Guid.NewGuid(), "Matcha", 30.00m, 2, new StarRating(5))
        };
    }

    private Mock<IInventoryQuery> GetMockQuery()
    {
        var mock = new Mock<IInventoryQuery>();
        mock.Setup(q => q.Execute()).Returns(_testItems.AsReadOnly());
        mock.Setup(q => q.AppliedFiltersAndSorts).Returns(new List<string>().AsReadOnly());
        return mock;
    }

    [Fact]
    public void AllInventoryQuery_Execute_ReturnsAllItems()
    {
        // Arrange
        var repository = new InventoryRepository();
        var query = new AllInventoryQuery(repository);

        // Act
        var result = query.Execute();

        // Assert
        Assert.Equal(repository.Get().Count, result.Count);
    }

    [Theory]
    [InlineData(true, 3)]
    [InlineData(false, 1)]
    [InlineData(null, 4)]
    public void AvailabilityFilter_Execute_FiltersCorrectly(bool? isAvailable, int expectedCount)
    {
        // Arrange
        var mockQuery = GetMockQuery();
        var filter = new AvailabilityFilterDecorator(mockQuery.Object, isAvailable);

        // Act
        var result = filter.Execute();

        // Assert
        Assert.Equal(expectedCount, result.Count);
    }

    [Fact]
    public void PriceRangeFilter_Execute_FiltersCorrectly()
    {
        // Arrange
        var mockQuery = GetMockQuery();
        var filter = new PriceRangeFilterDecorator(mockQuery.Object, 12.00m, 25.00m);

        // Act
        var result = filter.Execute();

        // Assert
        Assert.Equal(2, result.Count);
        Assert.All(result, i => Assert.True(i.Price >= 12.00m && i.Price <= 25.00m));
    }

    [Fact]
    public void NameContainsFilter_Execute_FiltersCorrectly()
    {
        // Arrange
        var mockQuery = GetMockQuery();
        var filter = new NameContainsFilterDecorator(mockQuery.Object, "Green");

        // Act
        var result = filter.Execute();

        // Assert
        Assert.Single(result);
        Assert.Equal("Green Tea", result[0].Name);
    }

    [Fact]
    public void SortByPrice_Execute_SortsDescendingCorrectly()
    {
        // Arrange
        var mockQuery = GetMockQuery();
        var sort = new SortByPriceDecorator(mockQuery.Object, SortDirection.Descending);

        // Act
        var result = sort.Execute();

        // Assert
        Assert.Equal(30.00m, result[0].Price);
        Assert.Equal(20.00m, result[1].Price);
        Assert.Equal(15.00m, result[2].Price);
        Assert.Equal(10.00m, result[3].Price);
    }

    [Fact]
    public void CombinedFiltersAndSorts_WorksCorrectly()
    {
        // Arrange
        var mockQuery = GetMockQuery();
        var availabilityFilter = new AvailabilityFilterDecorator(mockQuery.Object, true);
        var priceFilter = new PriceRangeFilterDecorator(availabilityFilter, 10.00m, 20.00m);
        var sort = new SortByPriceDecorator(priceFilter);

        // Act
        var result = sort.Execute();

        // Assert
        Assert.Equal(2, result.Count); // Green Tea (15), Oolong Tea (20)
        Assert.Equal("Green Tea", result[0].Name);
        Assert.Equal("Oolong Tea", result[1].Name);

        Assert.Equal(3, sort.AppliedFiltersAndSorts.Count);
    }
}

File: assignments/assignment-2-solution/src-java/src/test/java/teashop/unittest/domain/inventoryquery/QueryTests.java
package teashop.unittest.domain.inventoryquery;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventory.InventoryRepository;
import assignment2solution.domain.inventory.StarRating;
import assignment2solution.domain.inventoryquery.AllInventoryQuery;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.filters.AvailabilityFilterDecorator;
import assignment2solution.domain.inventoryquery.filters.NameContainsFilterDecorator;
import assignment2solution.domain.inventoryquery.filters.PriceRangeFilterDecorator;
import assignment2solution.domain.inventoryquery.sorts.SortByPriceDecorator;
import assignment2solution.domain.inventoryquery.sorts.SortDirection;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

public class QueryTests {
    private final List<InventoryItem> testItems = List.of(
        new InventoryItem(UUID.randomUUID(), "Green Tea", new BigDecimal("15.00"), 10, new StarRating(4)),
        new InventoryItem(UUID.randomUUID(), "Black Tea", new BigDecimal("10.00"), 0, new StarRating(5)),
        new InventoryItem(UUID.randomUUID(), "Oolong Tea", new BigDecimal("20.00"), 5, new StarRating(3)),
        new InventoryItem(UUID.randomUUID(), "Matcha", new BigDecimal("30.00"), 2, new StarRating(5))
    );

    private IInventoryQuery getStubQuery() {
        return new IInventoryQuery() {
            @Override
            public List<String> getAppliedFiltersAndSorts() {
                return List.of();
            }

            @Override
            public List<InventoryItem> execute() {
                return testItems;
            }
        };
    }

    @Test
    public void allInventoryQueryExecuteReturnsAllItems() {
        var repository = new InventoryRepository();
        var query = new AllInventoryQuery(repository);

        var result = query.execute();

        assertEquals(repository.get().size(), result.size());
    }

    @ParameterizedTest
    @CsvSource(value = {
        "true,3",
        "false,1",
        "null,4"
    }, nullValues = "null")
    public void availabilityFilterExecuteFiltersCorrectly(Boolean isAvailable, int expectedCount) {
        var mockQuery = getStubQuery();
        var filter = new AvailabilityFilterDecorator(mockQuery, isAvailable);

        var result = filter.execute();

        assertEquals(expectedCount, result.size());
    }

    @Test
    public void priceRangeFilterExecuteFiltersCorrectly() {
        var mockQuery = getStubQuery();
        var filter = new PriceRangeFilterDecorator(mockQuery, new BigDecimal("12.00"), new BigDecimal("25.00"));

        var result = filter.execute();

        assertEquals(2, result.size());
        assertTrue(result.stream().allMatch(i ->
            i.getPrice().compareTo(new BigDecimal("12.00")) >= 0
                && i.getPrice().compareTo(new BigDecimal("25.00")) <= 0));
    }

    @Test
    public void nameContainsFilterExecuteFiltersCorrectly() {
        var mockQuery = getStubQuery();
        var filter = new NameContainsFilterDecorator(mockQuery, "Green");

        var result = filter.execute();

        assertEquals(1, result.size());
        assertEquals("Green Tea", result.get(0).getName());
    }

    @Test
    public void sortByPriceExecuteSortsDescendingCorrectly() {
        var mockQuery = getStubQuery();
        var sort = new SortByPriceDecorator(mockQuery, SortDirection.DESCENDING);

        var result = sort.execute();

        assertEquals(new BigDecimal("30.00"), result.get(0).getPrice());
        assertEquals(new BigDecimal("20.00"), result.get(1).getPrice());
        assertEquals(new BigDecimal("15.00"), result.get(2).getPrice());
        assertEquals(new BigDecimal("10.00"), result.get(3).getPrice());
    }

    @Test
    public void combinedFiltersAndSortsWorksCorrectly() {
        var mockQuery = getStubQuery();
        var availabilityFilter = new AvailabilityFilterDecorator(mockQuery, true);
        var priceFilter = new PriceRangeFilterDecorator(availabilityFilter, new BigDecimal("10.00"), new BigDecimal("20.00"));
        var sort = new SortByPriceDecorator(priceFilter);

        var result = sort.execute();

        assertEquals(2, result.size());
        assertEquals("Green Tea", result.get(0).getName());
        assertEquals("Oolong Tea", result.get(1).getName());
        assertEquals(3, sort.getAppliedFiltersAndSorts().size());
    }
}
